// resolve-tspaths@0.8.18/dist/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const ansi_colors_1 = require("ansi-colors");
class Logger {
    constructor(level) {
        this.level = level;
    }
    verbose(...args) {
        if (this.level === "verbose") {
            console.log(...args);
        }
    }
    info(...args) {
        if (["verbose", "info"].includes(this.level)) {
            console.log(...args);
        }
    }
    error(...args) {
        console.error(...args.map((x) => (0, ansi_colors_1.red)(x)));
    }
    fancyParams(title, params) {
        this.verbose((0, ansi_colors_1.bold)(title));
        const keys = Object.keys(params);
        const isArray = Array.isArray(params);
        if (keys.length === 0)
            this.verbose((0, ansi_colors_1.dim)("empty"));
        else {
            for (const key of keys) {
                let value = params[key];
                if (typeof value === "string")
                    value = (0, ansi_colors_1.green)(value);
                if (isArray)
                    this.verbose(value);
                else
                    this.verbose(key, "->", value);
            }
        }
        this.verbose();
    }
    fancyError(title, message) {
        console.error();
        console.error(ansi_colors_1.red.bold(title));
        console.error(message);
        console.error();
    }
}
exports.Logger = Logger;
